File: bindings/clojure/src/proven/safe_json.clj
;; SPDX-License-Identifier: PMPL-1.0
;; SPDX-FileCopyrightText: 2025 Hyperpolymath

(ns proven.safe-json
  "Safe JSON parsing with depth and size limits."
  (:require [clojure.string :as str])
  (:import [java.io StringReader PushbackReader]))

(def ^:dynamic *max-depth* 64)
(def ^:dynamic *max-string-length* 1000000)
(def ^:dynamic *max-array-length* 10000)
(def ^:dynamic *max-object-keys* 10000)

(declare parse-value)

(defn- skip-whitespace [^PushbackReader reader]
  (loop []
    (let [char-code (.read reader)]
      (when (>= char-code 0)
        (let [character (char char-code)]
          (if (Character/isWhitespace character)
            (recur)
            (.unread reader char-code)))))))

(defn- peek-char [^PushbackReader reader]
  (let [char-code (.read reader)]
    (when (>= char-code 0)
      (.unread reader char-code)
      (char char-code))))

(defn- read-char [^PushbackReader reader]
  (let [char-code (.read reader)]
    (when (>= char-code 0)
      (char char-code))))

(defn- expect-char [^PushbackReader reader expected]
  (let [actual (read-char reader)]
    (when (not= actual expected)
      (throw (ex-info "Unexpected character" {:expected expected :actual actual})))))

(defn- parse-string [^PushbackReader reader]
  (expect-char reader \")
  (let [sb (StringBuilder.)]
    (loop []
      (let [character (read-char reader)]
        (when (nil? character)
          (throw (ex-info "Unterminated string" {})))
        (cond
          (= character \")
          (let [result (.toString sb)]
            (when (> (count result) *max-string-length*)
              (throw (ex-info "String too long" {:length (count result) :max *max-string-length*})))
            result)

          (= character \\)
          (let [escape-char (read-char reader)]
            (case escape-char
              \" (.append sb \")
              \\ (.append sb \\)
              \/ (.append sb \/)
              \b (.append sb \backspace)
              \f (.append sb \formfeed)
              \n (.append sb \newline)
              \r (.append sb \return)
              \t (.append sb \tab)
              \u (let [hex-chars (apply str (repeatedly 4 #(read-char reader)))
                       code-point (Integer/parseInt hex-chars 16)]
                   (.append sb (char code-point)))
              (throw (ex-info "Invalid escape sequence" {:char escape-char})))
            (recur))

          :else
          (do
            (.append sb character)
            (recur)))))))

(defn- parse-number [^PushbackReader reader]
  (let [sb (StringBuilder.)]
    (loop []
      (let [char-code (.read reader)]
        (if (< char-code 0)
          nil
          (let [character (char char-code)]
            (if (or (Character/isDigit character)
                    (= character \-)
                    (= character \+)
                    (= character \.)
                    (= character \e)
                    (= character \E))
              (do (.append sb character)
                  (recur))
              (.unread reader char-code))))))
    (let [num-str (.toString sb)]
      (if (str/includes? num-str ".")
        (Double/parseDouble num-str)
        (Long/parseLong num-str)))))

(defn- parse-array [^PushbackReader reader depth]
  (when (>= depth *max-depth*)
    (throw (ex-info "Maximum depth exceeded" {:depth depth :max *max-depth*})))
  (expect-char reader \[)
  (skip-whitespace reader)
  (if (= (peek-char reader) \])
    (do (read-char reader) [])
    (loop [elements []
           count 0]
      (when (>= count *max-array-length*)
        (throw (ex-info "Array too long" {:length count :max *max-array-length*})))
      (let [element (parse-value reader (inc depth))
            new-elements (conj elements element)]
        (skip-whitespace reader)
        (let [next-char (read-char reader)]
          (cond
            (= next-char \]) new-elements
            (= next-char \,) (do (skip-whitespace reader)
                                 (recur new-elements (inc count)))
            :else (throw (ex-info "Expected , or ]" {:char next-char}))))))))

(defn- parse-object [^PushbackReader reader depth]
  (when (>= depth *max-depth*)
    (throw (ex-info "Maximum depth exceeded" {:depth depth :max *max-depth*})))
  (expect-char reader \{)
  (skip-whitespace reader)
  (if (= (peek-char reader) \})
    (do (read-char reader) {})
    (loop [obj {}
           count 0]
      (when (>= count *max-object-keys*)
        (throw (ex-info "Object has too many keys" {:count count :max *max-object-keys*})))
      (skip-whitespace reader)
      (let [key (parse-string reader)]
        (skip-whitespace reader)
        (expect-char reader \:)
        (skip-whitespace reader)
        (let [value (parse-value reader (inc depth))
              new-obj (assoc obj key value)]
          (skip-whitespace reader)
          (let [next-char (read-char reader)]
            (cond
              (= next-char \}) new-obj
              (= next-char \,) (recur new-obj (inc count))
              :else (throw (ex-info "Expected , or }" {:char next-char})))))))))

(defn- parse-literal [^PushbackReader reader]
  (let [sb (StringBuilder.)]
    (loop []
      (let [char-code (.read reader)]
        (if (< char-code 0)
          nil
          (let [character (char char-code)]
            (if (Character/isLetter character)
              (do (.append sb character)
                  (recur))
              (.unread reader char-code))))))
    (let [literal (.toString sb)]
      (case literal
        "true" true
        "false" false
        "null" nil
        (throw (ex-info "Invalid literal" {:literal literal}))))))

(defn- parse-value [^PushbackReader reader depth]
  (skip-whitespace reader)
  (let [character (peek-char reader)]
    (cond
      (nil? character) (throw (ex-info "Unexpected end of input" {}))
      (= character \") (parse-string reader)
      (= character \{) (parse-object reader depth)
      (= character \[) (parse-array reader depth)
      (or (Character/isDigit character) (= character \-)) (parse-number reader)
      :else (parse-literal reader))))

(defn parse
  "Parse a JSON string.
   Returns {:ok value} or {:error message}.
   Options:
   - :max-depth - Maximum nesting depth (default 64)
   - :max-string-length - Maximum string length (default 1000000)
   - :max-array-length - Maximum array elements (default 10000)
   - :max-object-keys - Maximum object keys (default 10000)"
  ([json-string] (parse json-string {}))
  ([json-string options]
   (if (str/blank? json-string)
     {:error "empty_json"}
     (binding [*max-depth* (get options :max-depth *max-depth*)
               *max-string-length* (get options :max-string-length *max-string-length*)
               *max-array-length* (get options :max-array-length *max-array-length*)
               *max-object-keys* (get options :max-object-keys *max-object-keys*)]
       (try
         (let [reader (PushbackReader. (StringReader. json-string))
               result (parse-value reader 0)]
           (skip-whitespace reader)
           (if (>= (.read reader) 0)
             {:error "trailing_content"}
             {:ok result}))
         (catch Exception e
           {:error (str "parse_error: " (.getMessage e))}))))))

(defn valid?
  "Check if a string is valid JSON."
  [json-string]
  (contains? (parse json-string) :ok))

(defn parse!
  "Parse JSON string. Throws ex-info on error."
  ([json-string] (parse! json-string {}))
  ([json-string options]
   (let [result (parse json-string options)]
     (if (:ok result)
       (:ok result)
       (throw (ex-info (:error result) {:type :json-parse-error}))))))

(defn get-path
  "Get value at a path in parsed JSON. Path is a vector of keys/indices."
  [json-value path]
  (reduce (fn [current segment]
            (cond
              (nil? current) nil
              (and (map? current) (string? segment)) (get current segment)
              (and (vector? current) (integer? segment)) (get current segment)
              :else nil))
          json-value
          path))

(defn safe-get
  "Safely get a value with type checking."
  [json-value key expected-type]
  (let [value (get json-value key)]
    (when (instance? expected-type value)
      value)))

File: bindings/lua/proven/safe_json.lua
-- SPDX-License-Identifier: PMPL-1.0-or-later
-- SPDX-FileCopyrightText: 2025 Hyperpolymath

--- Safe JSON parsing and generation with depth limits.
--- @module proven.safe_json

local safe_json = {}

-- Default limits
local DEFAULT_MAX_DEPTH = 100
local DEFAULT_MAX_STRING_LENGTH = 1048576 -- 1MB
local DEFAULT_MAX_KEYS = 10000

--- Encode a Lua value as JSON.
--- @param value any The value to encode
--- @param options table|nil Optional settings (pretty, max_depth)
--- @return string|nil json The JSON string or nil on error
--- @return string|nil error Error message if encoding failed
function safe_json.encode(value, options)
    options = options or {}
    local max_depth = options.max_depth or DEFAULT_MAX_DEPTH
    local pretty = options.pretty or false
    local indent = options.indent or "  "

    local function encode_value(val, depth, current_indent)
        if depth > max_depth then
            return nil, "Maximum depth exceeded"
        end

        local val_type = type(val)

        if val == nil then
            return "null"
        elseif val_type == "boolean" then
            return val and "true" or "false"
        elseif val_type == "number" then
            if val ~= val then -- NaN check
                return "null"
            elseif val == math.huge or val == -math.huge then
                return "null"
            else
                return tostring(val)
            end
        elseif val_type == "string" then
            return safe_json.encode_string(val)
        elseif val_type == "table" then
            return safe_json.encode_table(val, depth, max_depth, pretty, indent, current_indent)
        else
            return nil, "Cannot encode type: " .. val_type
        end
    end

    return encode_value(value, 0, "")
end

--- Encode a string as a JSON string literal.
--- @param str string The string to encode
--- @return string The JSON string literal
function safe_json.encode_string(str)
    local escaped = str
        :gsub("\\", "\\\\")
        :gsub("\"", "\\\"")
        :gsub("\n", "\\n")
        :gsub("\r", "\\r")
        :gsub("\t", "\\t")
        :gsub("[\x00-\x1f]", function(c)
            return string.format("\\u%04x", string.byte(c))
        end)
    return '"' .. escaped .. '"'
end

--- Encode a table as JSON (array or object).
--- @param tbl table The table to encode
--- @param depth number Current depth
--- @param max_depth number Maximum allowed depth
--- @param pretty boolean Whether to format output
--- @param indent string Indentation string
--- @param current_indent string Current indentation level
--- @return string|nil json The JSON string or nil on error
--- @return string|nil error Error message if encoding failed
function safe_json.encode_table(tbl, depth, max_depth, pretty, indent, current_indent)
    if depth > max_depth then
        return nil, "Maximum depth exceeded"
    end

    -- Check if it's an array
    local is_array = true
    local max_index = 0
    local count = 0

    for k, _ in pairs(tbl) do
        count = count + 1
        if count > DEFAULT_MAX_KEYS then
            return nil, "Too many keys"
        end
        if type(k) ~= "number" or k < 1 or k ~= math.floor(k) then
            is_array = false
        else
            max_index = math.max(max_index, k)
        end
    end

    is_array = is_array and max_index == count

    local next_indent = current_indent .. indent
    local separator = pretty and ",\n" or ","
    local open_brace_suffix = pretty and "\n" or ""
    local close_brace_prefix = pretty and "\n" .. current_indent or ""
    local colon = pretty and ": " or ":"

    local parts = {}

    if is_array then
        for i = 1, max_index do
            local val = tbl[i]
            local encoded, err = safe_json.encode(val, {max_depth = max_depth - depth - 1, pretty = pretty, indent = indent})
            if not encoded then
                return nil, err
            end
            if pretty then
                table.insert(parts, next_indent .. encoded)
            else
                table.insert(parts, encoded)
            end
        end
        return "[" .. open_brace_suffix .. table.concat(parts, separator) .. close_brace_prefix .. "]"
    else
        for k, v in pairs(tbl) do
            local key_str = tostring(k)
            local encoded_value, err = safe_json.encode(v, {max_depth = max_depth - depth - 1, pretty = pretty, indent = indent})
            if not encoded_value then
                return nil, err
            end
            local pair = safe_json.encode_string(key_str) .. colon .. encoded_value
            if pretty then
                table.insert(parts, next_indent .. pair)
            else
                table.insert(parts, pair)
            end
        end
        table.sort(parts)
        return "{" .. open_brace_suffix .. table.concat(parts, separator) .. close_brace_prefix .. "}"
    end
end

--- Decode a JSON string to a Lua value.
--- @param json_string string The JSON string to decode
--- @param options table|nil Optional settings (max_depth, max_string_length)
--- @return any|nil value The decoded value or nil on error
--- @return string|nil error Error message if decoding failed
function safe_json.decode(json_string, options)
    if type(json_string) ~= "string" then
        return nil, "Input must be a string"
    end

    options = options or {}
    local max_depth = options.max_depth or DEFAULT_MAX_DEPTH
    local max_string_length = options.max_string_length or DEFAULT_MAX_STRING_LENGTH

    if #json_string > max_string_length then
        return nil, "JSON string too long"
    end

    local pos = 1
    local depth = 0

    local function skip_whitespace()
        while pos <= #json_string do
            local c = json_string:sub(pos, pos)
            if c ~= " " and c ~= "\t" and c ~= "\n" and c ~= "\r" then
                break
            end
            pos = pos + 1
        end
    end

    local function parse_value()
        skip_whitespace()

        if pos > #json_string then
            return nil, "Unexpected end of input"
        end

        local c = json_string:sub(pos, pos)

        if c == "n" then
            if json_string:sub(pos, pos + 3) == "null" then
                pos = pos + 4
                return nil
            end
        elseif c == "t" then
            if json_string:sub(pos, pos + 3) == "true" then
                pos = pos + 4
                return true
            end
        elseif c == "f" then
            if json_string:sub(pos, pos + 4) == "false" then
                pos = pos + 5
                return false
            end
        elseif c == '"' then
            return parse_string()
        elseif c == "[" then
            return parse_array()
        elseif c == "{" then
            return parse_object()
        elseif c == "-" or (c >= "0" and c <= "9") then
            return parse_number()
        end

        return nil, "Unexpected character: " .. c
    end

    local function parse_string()
        pos = pos + 1 -- Skip opening quote
        local start = pos
        local result = {}

        while pos <= #json_string do
            local c = json_string:sub(pos, pos)
            if c == '"' then
                pos = pos + 1
                return table.concat(result)
            elseif c == "\\" then
                pos = pos + 1
                if pos > #json_string then
                    return nil, "Unexpected end of string"
                end
                local escape = json_string:sub(pos, pos)
                if escape == "n" then
                    table.insert(result, "\n")
                elseif escape == "r" then
                    table.insert(result, "\r")
                elseif escape == "t" then
                    table.insert(result, "\t")
                elseif escape == '"' then
                    table.insert(result, '"')
                elseif escape == "\\" then
                    table.insert(result, "\\")
                elseif escape == "u" then
                    local hex = json_string:sub(pos + 1, pos + 4)
                    local code = tonumber(hex, 16)
                    if code then
                        if code < 128 then
                            table.insert(result, string.char(code))
                        else
                            -- Simple UTF-8 encoding
                            if code < 0x800 then
                                table.insert(result, string.char(0xC0 + math.floor(code / 64), 0x80 + (code % 64)))
                            else
                                table.insert(result, string.char(0xE0 + math.floor(code / 4096), 0x80 + math.floor((code % 4096) / 64), 0x80 + (code % 64)))
                            end
                        end
                        pos = pos + 4
                    end
                else
                    table.insert(result, escape)
                end
                pos = pos + 1
            else
                table.insert(result, c)
                pos = pos + 1
            end
        end

        return nil, "Unterminated string"
    end

    local function parse_number()
        local start = pos
        if json_string:sub(pos, pos) == "-" then
            pos = pos + 1
        end

        while pos <= #json_string and json_string:sub(pos, pos):match("[0-9]") do
            pos = pos + 1
        end

        if pos <= #json_string and json_string:sub(pos, pos) == "." then
            pos = pos + 1
            while pos <= #json_string and json_string:sub(pos, pos):match("[0-9]") do
                pos = pos + 1
            end
        end

        if pos <= #json_string and json_string:sub(pos, pos):lower() == "e" then
            pos = pos + 1
            if pos <= #json_string and json_string:sub(pos, pos):match("[+-]") then
                pos = pos + 1
            end
            while pos <= #json_string and json_string:sub(pos, pos):match("[0-9]") do
                pos = pos + 1
            end
        end

        local num_str = json_string:sub(start, pos - 1)
        return tonumber(num_str)
    end

    local function parse_array()
        depth = depth + 1
        if depth > max_depth then
            return nil, "Maximum depth exceeded"
        end

        pos = pos + 1 -- Skip [
        local arr = {}

        skip_whitespace()
        if json_string:sub(pos, pos) == "]" then
            pos = pos + 1
            depth = depth - 1
            return arr
        end

        while true do
            local val, err = parse_value()
            if err then
                return nil, err
            end
            table.insert(arr, val)

            skip_whitespace()
            local c = json_string:sub(pos, pos)
            if c == "]" then
                pos = pos + 1
                depth = depth - 1
                return arr
            elseif c == "," then
                pos = pos + 1
            else
                return nil, "Expected , or ]"
            end
        end
    end

    local function parse_object()
        depth = depth + 1
        if depth > max_depth then
            return nil, "Maximum depth exceeded"
        end

        pos = pos + 1 -- Skip {
        local obj = {}

        skip_whitespace()
        if json_string:sub(pos, pos) == "}" then
            pos = pos + 1
            depth = depth - 1
            return obj
        end

        while true do
            skip_whitespace()
            if json_string:sub(pos, pos) ~= '"' then
                return nil, "Expected string key"
            end

            local key, err = parse_string()
            if err then
                return nil, err
            end

            skip_whitespace()
            if json_string:sub(pos, pos) ~= ":" then
                return nil, "Expected :"
            end
            pos = pos + 1

            local val
            val, err = parse_value()
            if err then
                return nil, err
            end

            obj[key] = val

            skip_whitespace()
            local c = json_string:sub(pos, pos)
            if c == "}" then
                pos = pos + 1
                depth = depth - 1
                return obj
            elseif c == "," then
                pos = pos + 1
            else
                return nil, "Expected , or }"
            end
        end
    end

    local result, err = parse_value()
    if err then
        return nil, err
    end

    skip_whitespace()
    if pos <= #json_string then
        return nil, "Unexpected content after JSON value"
    end

    return result
end

--- Check if a string is valid JSON.
--- @param json_string string The string to validate
--- @return boolean valid Whether the string is valid JSON
function safe_json.is_valid(json_string)
    local _, err = safe_json.decode(json_string)
    return err == nil
end

--- Safely get a nested value from a decoded JSON object.
--- @param obj table The decoded JSON object
--- @param path string Dot-separated path (e.g., "user.name.first")
--- @return any|nil value The value at the path or nil if not found
function safe_json.get_path(obj, path)
    if type(obj) ~= "table" or type(path) ~= "string" then
        return nil
    end

    local current = obj
    for key in path:gmatch("[^%.]+") do
        if type(current) ~= "table" then
            return nil
        end
        local num_key = tonumber(key)
        current = current[num_key or key]
    end

    return current
end

return safe_json
